lib/posthog/feature_flags.rb

```diff
@@ -42,96 +77,288 @@ def is_feature_enabled(key, distinct_id, default_result = false, groups = {})
+          end
+        rescue InconclusiveMatchError => e
+          is_inconclusive = true
+        end
+      end
+
+      if !result.nil?
+        return result
+      elsif is_inconclusive
+        raise InconclusiveMatchError.new("Can't determine if feature flag is enabled or not with given properties")
+      end
 
-      decide_data = _request('POST', 'decide/?v=2', false)
-      feature_variants = decide_data["featureFlags"]
-      return feature_variants
+      # We can only return False when either all conditions are False, or
+      # no condition was inconclusive.
+      return false
     end
 
-    def get_feature_flag(key, distinct_id, groups = {})
-      variants = get_feature_variants(distinct_id, groups=groups)
-      return variants.fetch(key)
+    def is_condition_match(flag, distinct_id, condition, properties)
+      rollout_percentage = condition['rollout_percentage']
+
+      if !(condition['properties'] || []).empty?
+        if !condition['properties'].all? { |prop|
+            FeatureFlagsPoller.match_property(prop, properties)
+          }
+          return false
+        elsif rollout_percentage.nil?
+          return true
+        end
+      end
+
+      if !rollout_percentage.nil? and _hash(flag['key'], distinct_id) > (rollout_percentage.to_f/100)
+        return false
+      end
+      
+      return true
     end
 
-    def shutdown_poller()
-      @task.shutdown
+    # This function takes a distinct_id and a feature flag key and returns a float between 0 and 1.
+    # Given the same distinct_id and key, it'll always return the same float. These floats are
+    # uniformly distributed between 0 and 1, so if we want to show this feature to 20% of traffic
+    # we can do _hash(key, distinct_id) < 0.2
+    def _hash(key, distinct_id, salt="")
+      hash_key = Digest::SHA1.hexdigest "#{key}.#{distinct_id}#{salt}"
+      return (Integer(hash_key[0..14], 16).to_f / 0xfffffffffffffff)
     end
 
-    private
+    def get_matching_variant(flag, distinct_id)
+      hash_value = _hash(flag['key'], distinct_id, salt="variant")
+      variant_lookup_table(flag).each do |variant|
+        if (
+           hash_value >= variant[:value_min] and hash_value <= variant[:value_max]
+        )
+          return variant[:key]
+        end
+      end
+      return nil
+    end
 
-    def _load_feature_flags()
-      res = _request('GET', 'api/feature_flag', true)
-      @feature_flags.clear
-      @feature_flags = res['results'].filter { |flag| flag['active'] }
-      if @loaded_flags_successfully_once.false?
-        @loaded_flags_successfully_once.make_true
+    def variant_lookup_table(flag)
+      lookup_table = []
+      value_min = 0
+      multivariates = ((flag['filters'] || {})['multivariate'] || {})['variants'] || []
+      multivariates.each do |variant|
+        value_max = value_min + variant['rollout_percentage'].to_f / 100
+        lookup_table << {'value_min': value_min, 'value_max': value_max, 'key': variant['key']}
```

Comment:
In some other code you're using symbolized keys - you're bound to introduce bugs/confusion when working with both. Suggestion: `symbolize_keys` with data from API.